lib/pymedphys/_experimental/streamlit/apps/mosaiq_claude_chat/app.py

```diff
@@ -0,0 +1,178 @@
+# Copyright (C) 2024 Simon Biggs
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import json
+import os
+import pathlib
+
+import httpx
+import trio
+from anthropic import AsyncAnthropic
+from anyio.from_thread import BlockingPortal
+from pymedphys._imports import streamlit as st
+
+import pymedphys
+from pymedphys._ai.sql_agent.messages import (
+    Messages,
+    message_content_as_plain_text,
+    receive_user_messages_and_call_assistant_loop,
+    write_message,
+)
+from pymedphys._mosaiq.server_from_bak import start_mssql_docker_image_with_bak_restore
+
+from ._trio import get_streamlit_trio_portal
+
+USER = "user"
+
+ANTHROPIC_API_LIMIT = 2
+
+
+def main():
+    _key_handling()
+
+    portal = get_streamlit_trio_portal()
+    messages = _get_messages()
+    message_send_channel = _get_message_send_channel(portal=portal, messages=messages)
+
+    with st.sidebar:
+        _transcript_downloads(messages)
+
+        bak_filepath = (
+            pathlib.Path(
+                st.text_input(".bak file path", value="~/mosaiq-data/db-dump.bak")
+            )
+            .expanduser()
+            .resolve()
+        )
+        if st.button("Start demo MOSAIQ server from .bak file"):
+            start_mssql_docker_image_with_bak_restore(
+                bak_filepath=bak_filepath,
+                mssql_sa_password=os.getenv("MSSQL_SA_PASSWORD"),
+            )
+
+    for message in messages:
+        write_message(message["role"], message["content"])
+
+    new_message = st.chat_input()
+
+    if new_message:
+        portal.start_task_soon(
+            message_send_channel.send, {"role": USER, "content": new_message}
+        )
+
+    with st.sidebar:
+        st.write("---")
+        with st.spinner("Waiting forever for potential AI and tool use responses"):
+            portal.call(trio.sleep_forever)
+
+
+@st.cache_resource
+def _async_anthropic():
+    # TODO: Make this configurable
+    limits = httpx.Limits(max_connections=ANTHROPIC_API_LIMIT)
+
+    return AsyncAnthropic(connection_pool_limits=limits, max_retries=10)
+
+
+@st.cache_resource
+def _mosaiq_connection():
+    connection = pymedphys.mosaiq.connect(
+        "localhost",
+        database="PRACTICE",
+        username="sa",
+        password=os.environ["MSSQL_SA_PASSWORD"],
+    )
+    # Needed for multi-threading?
+    # https://stackoverflow.com/a/41912528
+    # connection._connection.autocommit(True)
```

Comment:
if the database is being accessed readonly, commits are something that shouldn't ever happen based on something from this connection.
What happens across queries that are not within the same query transaction if some other update takes place is going to be interesting.

Reply:
Yup... definitely interesting ...